src/main/java/io/cdap/plugin/gcp/datastore/source/DatastoreRecordReader.java

```diff
@@ -59,21 +62,33 @@ public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptCont
       config.getBoolean(DatastoreSourceConstants.CONFIG_SERVICE_ACCOUNT_IS_FILE, true),
       config.get(DatastoreSourceConstants.CONFIG_PROJECT));
     LOG.trace("Executing query split: {}", query);
-    RunQueryRequest request = RunQueryRequest.newBuilder()
-      .setQuery(query)
-      // partition id needs to be set in the RunQueryRequest in addition to being passed to QuerySplitter.getSplits.
-      // This is a quirk of the V1 API.
-      .setPartitionId(PartitionId.newBuilder()
-                        .setNamespaceId(config.get(DatastoreSourceConstants.CONFIG_NAMESPACE))
-                        .setProjectId(config.get(DatastoreSourceConstants.CONFIG_PROJECT)))
-      .build();
+
+    Query.Builder queryBuilder = query.toBuilder();
+    PartitionId.Builder partitionIdBuilder = PartitionId.newBuilder()
+      .setNamespaceId(config.get(DatastoreSourceConstants.CONFIG_NAMESPACE))
+      .setProjectId(config.get(DatastoreSourceConstants.CONFIG_PROJECT));
+    List<EntityResult> entityResultList = new ArrayList<>();
+    QueryResultBatch batch;
     try {
-      QueryResultBatch batch = datastore.runQuery(request).getBatch();
-      taskAttemptContext.getCounter(FileInputFormatCounter.BYTES_READ).increment(batch.getSerializedSize());
-      results = batch.getEntityResultsList().iterator();
+      // Datastore API only returns up to 300 items. Need to use cursor pagination if there is more results
+      do {
```

Comment:
Why this is done in the initialize method? Shouldn't reading of the actual data be done in the `nextKeyValue()` method?

Reply:
I added it here since the original `initialize()` was already loading in the first page of data, but I can see how this could be sub-optimal if there's a lot of records being loaded in. Would it be better to 1) not load any data in `initialize()` anymore (so the first page of records is pulled when `nextKeyValue()` is first called) or 2) load the first page of data when `initialize()` is called and load subsequent pages with `nextKeyValue()`?

Reply:
I think the latter will be better, 1st batch can done in initialize.